async_fuse/src/fuse/fuse_reply.rs

```diff
@@ -22,118 +20,154 @@ use super::protocol::{
-struct ReplyRaw<T: Send + Sync + 'static> {
+struct ReplyRaw {
     /// The FUSE request unique ID
     unique: u64,
     /// The FUSE device fd
     fd: RawFd,
-    /// The phantom data
-    marker: PhantomData<T>,
 }
 
-impl<T: Send + Sync + 'static> ReplyRaw<T> {
+impl ReplyRaw {
     /// Create `ReplyRaw`
     fn new(unique: u64, fd: RawFd) -> Self {
-        Self {
-            unique,
-            fd,
-            marker: PhantomData,
-        }
+        Self { unique, fd }
     }
 
     /// Send response to FUSE kernel
-    async fn send(self, to_bytes: ToBytes<T>, err: c_int) -> nix::Result<usize> {
+    async fn send(self, data: impl AsIoVecList + Send + Sync + 'static) -> nix::Result<usize> {
         let fd = self.fd;
         let wsize = smol::unblock(move || {
-            let mut send_error = false;
-            let instance: T; // to hold the instance of ToBytes::Struct
-            let byte_vec: Vec<u8>; // to hold the Vec<u8> of ToBytes::Bytes
-            let empty_vec: Vec<u8>; // to hold the emtpy Vec<u8> of ToBytes::Null
-            let (data_len, bytes) = match to_bytes {
-                ToBytes::Struct(inst) => {
-                    instance = inst;
-                    let len = mem::size_of::<T>();
-                    let bytes = match len {
-                        0 => &[],
-                        l => {
-                            let p = utilities::cast_to_ptr(&instance);
-                            unsafe { slice::from_raw_parts(p, l) }
-                        }
-                    };
-                    (len, bytes)
-                }
-                ToBytes::Bytes(bv) => {
-                    byte_vec = bv;
-                    (byte_vec.len(), &byte_vec[..])
-                }
-                ToBytes::Error => {
-                    send_error = true;
-                    empty_vec = Vec::new();
-                    (0, &empty_vec[..])
-                }
-            };
             let header_len = mem::size_of::<FuseOutHeader>();
+
             let header = FuseOutHeader {
-                len: (header_len.overflow_add(data_len)).cast(),
-                error: err.overflow_neg(), // FUSE requires the error number to be negative
+                len: (header_len.overflow_add(data.len())).cast(),
+                error: 0, // FUSE requires the error number to be negative
                 unique: self.unique,
             };
+
             let h = utilities::cast_to_ptr(&header);
             let header_bytes = unsafe { slice::from_raw_parts(h, header_len) };
-            let iovecs: Vec<_> = if data_len > 0 {
-                vec![IoVec::from_slice(header_bytes), IoVec::from_slice(bytes)]
+            let iovecs: Vec<_> = if data.len() > 0 {
+                let mut header = vec![IoVec::from_slice(header_bytes)];
+                header.append(&mut data.as_io_vec_list());
+                header
             } else {
                 vec![IoVec::from_slice(header_bytes)]
             };
```

Comment:
The latter allocation can be reduced by using on-stack DST coercion.
Here is an example:
```rust
let cond = false;

let (v1, v2);
let s: &[u8] = if cond {
    v1 = vec![1, 2];
    &*v1
} else {
    v2 = [3];
    &v2
};

println!("{:?}", s);
```